File: DataEngineering/02_Data_Modeling/projects/python/03_data_vault_hub_link_sat/main.py
import hashlib
import logging
from datetime import datetime
from sqlalchemy import create_engine, Column, String, DateTime, ForeignKey, PrimaryKeyConstraint
from sqlalchemy.orm import declarative_base, sessionmaker, relationship

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

Base = declarative_base()

def hash_key(*args):
    """Generates an MD5 hash key for business keys."""
    combined = "|".join([str(arg) for arg in args])
    return hashlib.md5(combined.encode('utf-8')).hexdigest()

# ----------------- HUBS -----------------
class HubCustomer(Base):
    __tablename__ = 'hub_customer'
    customer_hk = Column(String(32), primary_key=True)
    customer_id = Column(String(50), nullable=False, unique=True)
    load_date = Column(DateTime, nullable=False)
    record_source = Column(String(50), nullable=False)

class HubOrder(Base):
    __tablename__ = 'hub_order'
    order_hk = Column(String(32), primary_key=True)
    order_id = Column(String(50), nullable=False, unique=True)
    load_date = Column(DateTime, nullable=False)
    record_source = Column(String(50), nullable=False)

# ----------------- LINKS -----------------
class LinkCustomerOrder(Base):
    __tablename__ = 'link_customer_order'
    link_hk = Column(String(32), primary_key=True)
    customer_hk = Column(String(32), ForeignKey('hub_customer.customer_hk'), nullable=False)
    order_hk = Column(String(32), ForeignKey('hub_order.order_hk'), nullable=False)
    load_date = Column(DateTime, nullable=False)
    record_source = Column(String(50), nullable=False)

    customer = relationship("HubCustomer")
    order = relationship("HubOrder")

# --------------- SATELLITES ---------------
class SatCustomerDetails(Base):
    __tablename__ = 'sat_customer_details'
    customer_hk = Column(String(32), ForeignKey('hub_customer.customer_hk'), nullable=False)
    load_date = Column(DateTime, nullable=False)
    name = Column(String(100), nullable=False)
    email = Column(String(150), nullable=False)
    record_source = Column(String(50), nullable=False)

    # Composite primary key for Satellite to track history
    __table_args__ = (
        PrimaryKeyConstraint('customer_hk', 'load_date'),
    )

    customer = relationship("HubCustomer")

def main():
    # Production Data Vaults are typically on Snowflake, Redshift, or Postgres.
    DATABASE_URL = "sqlite:///:memory:"
    engine = create_engine(DATABASE_URL, echo=False)
    Base.metadata.create_all(engine)
    
    Session = sessionmaker(bind=engine)
    session = Session()

    try:
        ld = datetime.now()
        
        # 1. Insert Hub & Sat for Customer
        cust_id = 'C123'
        c_hk = hash_key(cust_id)
        
        hub_c = HubCustomer(customer_hk=c_hk, customer_id=cust_id, load_date=ld, record_source='CRM_SYS')
        sat_c = SatCustomerDetails(customer_hk=c_hk, load_date=ld, name='Alice Smith', email='alice@example.com', record_source='CRM_SYS')
        session.add_all([hub_c, sat_c])

        # 2. Insert Hub for Order
        ord_id = 'O999'
        o_hk = hash_key(ord_id)
        
        hub_o = HubOrder(order_hk=o_hk, order_id=ord_id, load_date=ld, record_source='ORDER_SYS')
        session.add(hub_o)

        # 3. Insert Link connecting Customer and Order
        l_hk = hash_key(c_hk, o_hk)
        link_co = LinkCustomerOrder(link_hk=l_hk, customer_hk=c_hk, order_hk=o_hk, load_date=ld, record_source='ORDER_SYS')
        session.add(link_co)

        session.commit()
        logger.info("Data Vault Architecture Initialized & Data Inserted.")

        # Reconstruct Business View
        logger.info("Querying Data Vault to reconstruct business view:")
        results = session.query(
            HubCustomer.customer_id,
            SatCustomerDetails.name,
            HubOrder.order_id
        ).select_from(LinkCustomerOrder) \
         .join(HubCustomer, LinkCustomerOrder.customer_hk == HubCustomer.customer_hk) \
         .join(HubOrder, LinkCustomerOrder.order_hk == HubOrder.order_hk) \
         .join(SatCustomerDetails, HubCustomer.customer_hk == SatCustomerDetails.customer_hk) \
         .all()

        for r in results:
            logger.info(f"Customer ID: {r.customer_id} | Name: {r.name} | Order ID: {r.order_id}")

    except Exception as e:
        session.rollback()
        logger.error(f"Error in Data Vault ETL: {e}")
    finally:
        session.close()

if __name__ == '__main__':
    main()

File: DataEngineering/02_Data_Modeling/projects/golang/03_data_vault_hub_link_sat/main.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"log"
	"time"

	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"
)

// hashKey generates an MD5 hash key for Data Vault business keys
func hashKey(val string) string {
	hash := md5.Sum([]byte(val))
	return hex.EncodeToString(hash[:])
}

// ---------------- HUBs ----------------

type HubCustomer struct {
	HkCustomerID        string    `gorm:"primaryKey;type:varchar(32)"`
	CustomerBusinessKey string    `gorm:"not null;unique"`
	LoadDate            time.Time `gorm:"not null"`
	RecordSource        string    `gorm:"not null"`
}

type HubTransaction struct {
	HkTransactionID        string    `gorm:"primaryKey;type:varchar(32)"`
	TransactionBusinessKey string    `gorm:"not null;unique"`
	LoadDate               time.Time `gorm:"not null"`
	RecordSource           string    `gorm:"not null"`
}

// ---------------- LINKs ----------------

type LinkCustomerTransaction struct {
	HkLinkID        string    `gorm:"primaryKey;type:varchar(32)"`
	HkCustomerID    string    `gorm:"not null"`
	HkTransactionID string    `gorm:"not null"`
	LoadDate        time.Time `gorm:"not null"`
	RecordSource    string    `gorm:"not null"`

	Customer    HubCustomer    `gorm:"foreignKey:HkCustomerID"`
	Transaction HubTransaction `gorm:"foreignKey:HkTransactionID"`
}

// -------------- SATELLITEs --------------

type SatCustomerDetails struct {
	HkCustomerID string    `gorm:"primaryKey;type:varchar(32)"` // Composite PK part 1
	LoadDate     time.Time `gorm:"primaryKey"`                  // Composite PK part 2
	FirstName    string    `gorm:"not null"`
	LastName     string    `gorm:"not null"`
	RecordSource string    `gorm:"not null"`

	Customer HubCustomer `gorm:"foreignKey:HkCustomerID"`
}

func main() {
	dsn := "host=localhost user=postgres password=postgres dbname=data_vault port=5432 sslmode=disable"
	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{
		Logger: logger.Default.LogMode(logger.Info),
	})
	if err != nil {
		log.Printf("Failed to connect to database: %v", err)
		log.Println("Skipping actual execution...")
		return
	}

	// 1. Auto Migrate Data Vault Schema
	err = db.AutoMigrate(
		&HubCustomer{},
		&HubTransaction{},
		&LinkCustomerTransaction{},
		&SatCustomerDetails{},
	)
	if err != nil {
		log.Fatalf("Error creating Data Vault schema: %v", err)
	}
	fmt.Println("Data Vault Schema successfully initialized!")

	// 2. Load Data Vault Entities
	loadDate := time.Now()

	// Insert Hub & Satellite for Customer
	cID := "CUST-1001"
	hkCust := hashKey(cID)

	hubC := HubCustomer{
		HkCustomerID:        hkCust,
		CustomerBusinessKey: cID,
		LoadDate:            loadDate,
		RecordSource:        "CRM",
	}
	satC := SatCustomerDetails{
		HkCustomerID: hkCust,
		LoadDate:     loadDate,
		FirstName:    "Alice",
		LastName:     "Smith",
		RecordSource: "CRM",
	}
	db.Create(&hubC)
	db.Create(&satC)

	// Insert Hub for Transaction
	tID := "TX-9999"
	hkTx := hashKey(tID)

	hubT := HubTransaction{
		HkTransactionID:        hkTx,
		TransactionBusinessKey: tID,
		LoadDate:               loadDate,
		RecordSource:           "BILLING",
	}
	db.Create(&hubT)

	// Insert Link connecting Customer to Transaction
	hkLink := hashKey(hkCust + hkTx)
	link := LinkCustomerTransaction{
		HkLinkID:        hkLink,
		HkCustomerID:    hkCust,
		HkTransactionID: hkTx,
		LoadDate:        loadDate,
		RecordSource:    "BILLING",
	}
	db.Create(&link)
	fmt.Println("Data inserted into Data Vault.")

	// 3. Query the vault (Reconstruct business view)
	type BusinessView struct {
		CustomerKey string
		FirstName   string
		LastName    string
		TxKey       string
	}
	var results []BusinessView

	db.Table("link_customer_transactions").
		Select(`
			hub_customers.customer_business_key as customer_key,
			sat_customer_details.first_name,
			sat_customer_details.last_name,
			hub_transactions.transaction_business_key as tx_key
		`).
		Joins("JOIN hub_customers ON hub_customers.hk_customer_id = link_customer_transactions.hk_customer_id").
		Joins("JOIN hub_transactions ON hub_transactions.hk_transaction_id = link_customer_transactions.hk_transaction_id").
		Joins("JOIN sat_customer_details ON sat_customer_details.hk_customer_id = hub_customers.hk_customer_id").
		Scan(&results)

	fmt.Println("--- Business View ---")
	for _, r := range results {
		fmt.Printf("Customer: %s (%s %s) | Transaction: %s\n", r.CustomerKey, r.FirstName, r.LastName, r.TxKey)
	}
}
